arch/s390/dfltcc_deflate.c

```diff
@@ -101,31 +105,38 @@ int ZLIB_INTERNAL dfltcc_deflate(PREFIX3(streamp) strm, int flush, block_state *
     int soft_bcc;
     int no_flush;
 
-    if (!dfltcc_can_deflate(strm))
+    if (!dfltcc_can_deflate(strm)) {
+        /* Clear history. */
+        if (flush == Z_FULL_FLUSH)
+            param->hl = 0;
         return 0;
+    }
 
 again:
     masked_avail_in = 0;
     soft_bcc = 0;
     no_flush = flush == Z_NO_FLUSH;
 
-    /* Trailing empty block. Switch to software, except when Continuation Flag
-     * is set, which means that DFLTCC has buffered some output in the
-     * parameter block and needs to be called again in order to flush it.
+    /* No input data. Return, except when Continuation Flag is set, which means
+     * that DFLTCC has buffered some output in the parameter block and needs to
+     * be called again in order to flush it.
      */
-    if (flush == Z_FINISH && strm->avail_in == 0 && !param->cf) {
-        if (param->bcf) {
-            /* A block is still open, and the hardware does not support closing
-             * blocks without adding data. Thus, close it manually.
-             */
+    if (strm->avail_in == 0 && !param->cf) {
+        /* A block is still open, and the hardware does not support closing
```

Comment:
Deflate structure has a property called `block_open` . Can `param->cf` be replaced with it? Currently it is only used for `deflate_quick`.

Reply:
`param->cf` (Continuation Flag) is more like "did hardware buffer some output bits". `param->bcf`(Block Continuation Flag) is close to `block_open`, but it's set and read by hardware, so no full replacement is possible. However, it might be worth considering syncing the two before and after DFLTCC call.

Reply:
Btw, the instruction is finally publicly documented (this wasn't the case when I first sent the patches). Here is our latest official SystemZ architecture document - http://publibfp.dhe.ibm.com/epubs/pdf/a227832c.pdf. DFLTCC is in Chapter 26 (page 26-16).

Reply:
I have to correct myself: BCF is only read (and never set) by the hardware. I must have been thinking about CF when writing this. Still, the main point holds: since it's a hardware-defined bit, it cannot be fully thrown away.